### DISCUSSION TITLE
a REPL for development

### DISCUSSION THREAD

**Discussion by justin2004** [saulpw/visidata#2290] [Ideas]
Hi all,

I am making some new commands and I kept finding myself doing this:
```python
# BaseSheet.addCommand('1', 'hello-world', 'status(dir(sheet))')
BaseSheet.addCommand('1', 'hello-world', 'status(sheet.cursorCell.text)')
```
that is, using the status window as a console log to see what functions/attributes objects have.
But that requires that I save `.visidatarc` and re-run `vd`.

I really wanted a REPL for this so I put this into my `.visidatarc`.
```python
BaseSheet.addCommand('', 'REPL', """
import subprocess
import pprint
with SuspendCurses():
    pp = pprint.PrettyPrinter(indent=4).pprint
    # now you can `pp(dir(sheet))` for example
    while True:
        try:
            # print(eval(input(">>> "))) # this works if you don't want expression history
            print(eval(subprocess.run(['rlwrap','--one-shot','--substitute-prompt',">>> ",'cat'], text=True, capture_output=True).stdout))
        except Exception as e:
            # print(f"Error: {e}")
            print(e)
            break
""")
```

You'll first want to `apt-get install rlwrap` and if you like vim I would put this in `~/.inputrc`:
```
set editing-mode vi
```

Now I can inspect stuff at the REPL and the jump back into vd at the sheet where I left off.
e.g.
from a sheet I do `<space>REPL<enter>`
![image](https://github.com/saulpw/visidata/assets/16325158/465fbece-454b-4ef9-a430-affcc4b490c7)
![Screenshot from 2024-02-02 14-09-26](https://github.com/saulpw/visidata/assets/16325158/800b55fd-804e-4906-a176-8f31c088f7e7)

more usefully:
![image](https://github.com/saulpw/visidata/assets/16325158/248356ac-713d-41a7-ae2c-09e46d01d727)
![image](https://github.com/saulpw/visidata/assets/16325158/aa22087c-64fe-4113-a2c0-b35ba1f22d64)

I'd like to fix the exception handling:
![image](https://github.com/saulpw/visidata/assets/16325158/c0eea8e4-aea2-4758-8d0b-354ed5f7bc52)

Ideally we'd catch the EOF and return quietly and for syntax errors i'd like to `continue` and stay in the REPL.
Can anyone see the problem with that?

Well I hope this helps someone else too!

Note, I know about the `exec-python` command but you don't get command/expression history and it doesn't feel REPLy.

**Comment by ajkerrigan** [comment]
Using an rlwrap subprocess is a clever idea, I dig it!

I also found myself often wanting a REPL which led to the [vdrepl plugin](https://github.com/ajkerrigan/visidata-plugins/blob/main/plugins/vdrepl.py). That's a different approach with different assumptions/dependencies though. I haven't touched it because it still works, but it could probably use a cleanup pass.

Your method seems to work smoothly out of the box for me :muscle: . Regarding the `name 'e' is not defined` error, I know scoping can get pretty funky with the timing of evals and stuff. A couple tweaks like this seem to help dodge that issue, allow you to catch/display "real" errors, and silently exit back to vd on ^D or ^C:

```python
import subprocess
import pprint

def rlwrap_repl():
    with SuspendCurses():
        pp = pprint.PrettyPrinter(indent=4).pprint
        while True:
            try:
                source = subprocess.run(['rlwrap','--one-shot','--substitute-prompt',">>> ",'cat'], text=True, capture_output=True).stdout
                if not source:
                    break
                print(eval(source))
            except (EOFError, KeyboardInterrupt):
                break
            except Exception as e:
                print(e)
                continue

BaseSheet.addCommand('', 'REPL', 'rlwrap_repl()')
```

**Reply by justin2004** [reply]
thanks, @ajkerrigan !

when i was looking for a vd repl i didn't find your plugin. i'll try it out too!

**Comment by saulpw** [comment]
Since multiple people want a REPL for VisiData like this, what do y'all think about combining the best ideas from @justin2004's code and @ajkerrigan's vdrepl plugin and pulling them into the vd repo formally?

**Reply by justin2004** [reply]
@saulpw , sounds good!

note that i had to update mine to allow statements (like assignment):
```python
            except SyntaxError as e:
                # `source` might have been a statement (like assignment)
                exec(source)
```

**Reply by ajkerrigan** [reply]
Sounds good to me! Not sure the best way to combine things, but after using both side by side for a bit I realized my take wasn't bringing as many variables into the local scope. Tweaked that.

Some other thoughts:

- I think @justin2004 's implementation feels closer to vd's normal `exec-python` semantics, where ptipython might be heavier than most people want.
- rlwrap seems quick/easy to install across multiple platforms via apt/brew/etc :+1: . my approach leans on ptpython/ptipython being installed via pip. From the vd maintenance perspective does one fit better than the other? Feels pretty user-dependent to me, both totally valid.
- A formal repl should probably offer a history file option, but default to something inside the visidata dir (`~/.visidata/cache/repl_history` or something?)
  - After writing this reply I came across https://github.com/saulpw/visidata/issues/2298#issue-2117829654 which seems relevant for this

**Comment by saulpw** [comment]
BTW did y'all know about the `Ctrl+Y` family of commands?  They pull up a sheet with the row/cell/sheet of the attributes of the underlying Python object.

**Comment by frosencrantz** [comment]
@saulpw what happened with this feature:?  It looks like it did not get added this year.  I can add a feature request to track this, if that would help any.
